### ISSUE TITLE
`@Ignore` annotation on classes with JUnit 4's `Enclosed` runner causes deadlock

### ISSUE THREAD

**Issue by rschmitt** [gradle/gradle#8544]
### Expected Behavior
JUnit 4.12 includes an experimental test runner, `org.junit.experimental.runners.Enclosed`, which is similar to `@Nested` in JUnit 5. I've inherited a code base where this pattern is common:

```java
@RunWith(Enclosed.class)
public class BDDStyleTests {
    @Ignore
    public static class TestContextBase {
        public TestContextBase() {
            // Common setup goes here
        }
    }

    public static class StaticInnerTestClass extends TestContextBase {
        @Test
        public void testMethod() {
            // test method
        }
    }
}
```

The `TestContextBase` must be annotated with `@Ignore`, since it contains no test methods. Gradle does not seem to be tolerating this annotation well.

### Current Behavior

Using JUnit 4, I get the following:

```
Unable to load JUnit4 runner to calculate Ignored test cases
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at org.gradle.api.internal.tasks.testing.junit.AllExceptIgnoredTestRunnerBuilder$FallbackJUnit4Builder.runnerForClass(AllExceptIgnoredTestRunnerBuilder.java:54)
        at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
        at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
        at org.gradle.api.internal.tasks.testing.junit.IgnoredTestDescriptorProvider.getAllDescriptions(IgnoredTestDescriptorProvider.java:31)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestEventAdapter.processIgnoredClass(JUnitTestEventAdapter.java:106)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestEventAdapter.testIgnored(JUnitTestEventAdapter.java:95)
        at org.junit.runner.notification.SynchronizedRunListener.testIgnored(SynchronizedRunListener.java:77)
        at org.junit.runner.notification.RunNotifier$6.notifyListener(RunNotifier.java:174)
        at org.junit.runner.notification.RunNotifier$SafeNotifier.run(RunNotifier.java:72)
        at org.junit.runner.notification.RunNotifier.fireTestIgnored(RunNotifier.java:171)
        at org.junit.internal.builders.IgnoredClassRunner.run(IgnoredClassRunner.java:16)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
        at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
        at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
        at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
        at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:118)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:175)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:157)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.junit.internal.runners.InitializationError
        at org.junit.internal.runners.MethodValidator.assertValid(MethodValidator.java:57)
        at org.junit.internal.runners.JUnit4ClassRunner.validate(JUnit4ClassRunner.java:45)
        at org.junit.internal.runners.JUnit4ClassRunner.<init>(JUnit4ClassRunner.java:35)
        ... 53 more

BUILD SUCCESSFUL in 1s
```

The build does not fail, and the tests appear to run, but I'm still getting an exception (whose [original cause is being discarded](https://github.com/gradle/gradle/blob/master/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/AllExceptIgnoredTestRunnerBuilder.java#L49)).

If I use JUnit 5's Vintage runner and the JUnit Platform, the entire build locks up. It gets stuck at the class that contains the `@Ignore` annotation, and it stays locked until I kill the JVM.

### Context
I'm trying to migrate from Maven, which has had no issues with any version of JUnit.

### Steps to Reproduce (for bugs)
[repro.tar.gz](https://github.com/gradle/gradle/files/2877707/repro.tar.gz)

Extract and run `./gradlew test`. To reproduce the hanging issue with JUnit Platform, comment out the JUnit 4 stuff in build.gradle and uncomment the JUnit 5 stuff.

### Your Environment
```
------------------------------------------------------------
Gradle 5.2.1
------------------------------------------------------------

Build time:   2019-02-08 19:00:10 UTC
Revision:     f02764e074c32ee8851a4e1877dd1fea8ffb7183

Kotlin DSL:   1.1.3
Kotlin:       1.3.20
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          11 (Oracle Corporation 11+28)
OS:           Mac OS X 10.13.6 x86_64
```

I've also reproduced this bug on JDK8.

**Comment by rschmitt** [comment]
I've discovered a workaround, which is to simply declare the base class `abstract` instead of using the `@Ignore` annotation. I've also learned that on JUnit 4, the `BlockJUnit4ClassRunner` constructor is ultimately throwing an exception because the base class contains to runnable methods (in other words, it is apparently unaware of `@Ignore` annotations).

I have not investigated the problems with JUnitPlatform/JUnit 5 Vintage, but in my opinion the deadlock when using that runner is by far the most worrisome of the problems I'm seeing, and suggests a deficiency in error handling somewhere essential.
